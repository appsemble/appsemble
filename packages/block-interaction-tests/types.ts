declare module '@appsemble/sdk' {
  // eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
  interface EventEmitters {
    [K: string]: never;
  }

  // eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
  interface EventListeners {
    [K: string]: never;
  }
}
