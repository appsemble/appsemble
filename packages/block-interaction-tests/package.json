{
  "name": "@appsemble/block-interaction-tests",
  "version": "0.34.22-test.3",
  "private": true,
  "description": "Used to test interactions between different appsemble blocks.",
  "keywords": [
    "app",
    "apps",
    "framework",
    "low-code",
    "lowcode"
  ],
  "homepage": "https://appsemble.app",
  "bugs": "https://gitlab.com/appsemble/appsemble/-/issues",
  "repository": {
    "type": "git",
    "url": "https://gitlab.com/appsemble/appsemble.git",
    "directory": "packages/block-interaction-tests"
  },
  "license": "LGPL-3.0-only",
  "author": "Appsemble <info@appsemble.com> (https://appsemble.com)",
  "type": "module",
  "scripts": {
    "test": "vitest"
  },
  "dependencies": {
    "@appsemble/data-loader": "0.34.22-test.3",
    "@appsemble/list": "0.34.22-test.3",
    "@appsemble/preact": "0.34.22-test.3",
    "@appsemble/preact-components": "0.34.22-test.3",
    "@appsemble/sdk": "0.34.22-test.3",
    "@appsemble/utils": "0.34.22-test.3",
    "@types/node": "22.15.21",
    "classnames": "^2.0.0",
    "preact": "^10.0.0"
  },
  "devDependencies": {
    "@testing-library/preact": "3.2.4",
    "@testing-library/user-event": "14.6.1",
    "vitest": "2.1.9"
  }
}
