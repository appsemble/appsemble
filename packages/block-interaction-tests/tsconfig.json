{
  "extends": "../../tsconfig",
  "compilerOptions": {
    "rootDir": ".",
    "strictFunctionTypes": false,
    "strictNullChecks": true,
    "types": ["@appsemble/webpack-config/types"],
    "lib": ["dom", "dom.iterable", "esnext"],
    "jsx": "react-jsx",
    "jsxImportSource": "preact",
    "moduleDetection": "force",
    "moduleResolution": "bundler",
    "module": "ESNext"
  }
}
