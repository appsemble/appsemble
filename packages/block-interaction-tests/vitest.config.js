import { createVitestConfig } from '../../config/vitest/config.js';

export default createVitestConfig(import.meta);
