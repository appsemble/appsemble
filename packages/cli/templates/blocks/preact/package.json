{
  "private": true,
  "type": "module",
  "dependencies": {
    "@appsemble/preact": "0.23.4",
    "@appsemble/sdk": "0.23.4",
    "preact": "^10.0.0"
  }
}
