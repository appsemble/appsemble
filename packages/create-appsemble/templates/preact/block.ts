declare module '@appsemble/sdk' {
  interface EventListeners {
    /**
     * This event can be used to receive incoming data to display.
     */
    data: never;
  }

  interface Parameters {
    /**
     * A list of fields to render out in a table.
     */
    fields: string[];
  }
}

export {};
