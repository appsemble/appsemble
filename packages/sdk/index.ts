import {
  type Action,
  type ActionError,
  type BaseMessage,
  type BulmaColor,
  type BulmaSize,
  type Theme,
} from '@appsemble/types';
import { type IconName } from '@fortawesome/fontawesome-common-types';
import { type Promisable } from 'type-fest';

export { type IconName };
export { type Action, type BulmaColor, type BulmaSize, type Theme };

/**
 * A user defined remapper function.
 */
export type Remapper = object[] | boolean | number | object | string;

/**
 * Actions defined on a block.
 *
 * If a block uses actions, extend this interface using module augmentation. The keys are the names
 * of the events the block supports.
 *
 * @example
 * declare module '@appsemble/sdk' {
 *   interface Actions {
 *     onClick: {}
 *   }
 * }
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface Actions {}

/**
 * Messages defined on a block.
 *
 * If a block uses messages, extend this interface using module augmentation.
 * The keys are the names of the messages.
 * The types are either `never` if the message doesn’t support parameters
 * or an object containing the keys of values and the type of the value.
 *
 * @example
 * declare module '@appsemble/sdk' {
 *   interface Messages {
 *     exampleMessage: never;
 *     hello: { person: string };
 *   }
 * }
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface Messages {}

/**
 * Event emitters defined on a block.
 *
 * If a block emits events, extend this interface using module augmentation. The keys are the names
 * of the events the block can emit.
 *
 * @example
 * declare module '@appsemble/sdk' {
 *   interface EventEmitters {
 *     data: {}
 *   }
 * }
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface EventEmitters {}

/**
 * Event listeners defined on a block.
 *
 * If a block listens on events, extend this interface using module augmentation. The keys are the
 * names of the events the block can emit.
 *
 * @example
 * declare module '@appsemble/sdk' {
 *   interface EventListeners {
 *     data: {}
 *   }
 * }
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface EventListeners {}

/**
 * Custom free form parameters defined on a block.
 *
 * If a block listens on events, extend this interface using module augmentation. The keys are the
 * names of the events the block can emit.
 *
 * @example
 * declare module '@appsemble/sdk' {
 *   interface Parameters {
 *     param1: string;
 *     param2: number;
 *     param3: {
 *       nested: boolean;
 *     }
 *   }
 * }
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface Parameters {}

export type PageParameters = Record<string, string>;

/**
 * A menu item that can be displayed in addition to the normal app menu.
 */
export interface MenuItem {
  /**
   * The title of the menu item.
   */
  title: string;

  /**
   * The icon to display next to the title.
   */
  icon?: IconName;

  /**
   * The color to use for the icon.
   */
  iconColor?: BulmaColor;

  /**
   * The click handler for when the menu item is clicked.
   */
  onClick: () => void;

  /**
   * The submenu items associated with this menu item.
   */
  submenu?: Omit<MenuItem, 'submenu'>[];

  /**
   * Whether the item is considered active.
   */
  active?: boolean;
}

export interface Message extends BaseMessage {
  /**
   * The content of the message to display.
   */
  body: string;
}

/**
 * A set of utility functions provided by the Appsemble SDK.
 */
export interface Utils {
  /**
   * Register a function that should be called when the block is being removed.
   *
   * Use this to clean up resouces that would otherwise stay in memory, e.g. object URLs.
   */
  addCleanup: (fn: () => void) => void;

  /**
   * Test if the input is an action error.
   *
   * @param input The input to test
   * @returns Whether or not the input is an action error.
   */
  isActionError: (input: unknown) => input is ActionError;

  /**
   * Remap data based in a user defined remapper function.
   *
   * @param remapper The user defined remapper function.
   * @param data The data to remap.
   * @param context The context to retrieve contextual data from.
   * @returns The result of the remapped data.
   */
  remap: (
    remapper: Remapper | undefined,
    data: unknown,
    context?: Record<string, unknown>,
  ) => unknown;
}

export interface BlockUtils extends Utils {
  /**
   * Show a bulma style message.
   */
  showMessage: (message: Message | string) => void;

  /**
   * Formats a message using ICU syntax.
   */
  formatMessage: <T extends keyof Messages>(
    message: T,
    ...args: Messages[T] extends never ? [] : [Messages[T]]
  ) => string;

  /**
   * Get a URL serving an asset for the given asset id.
   *
   * @param assetId The id of the asset whose URL to get.
   * @returns The URL that matches the given asset id.
   */
  asset: (assetId: string) => string;

  /**
   * Convert a FontAwesome icon name into a valid FontAwesome CSS class.
   *
   * @param icon The FontAwesome icon.
   * @returns String containing the FontAwesome classes for the icon.
   */
  fa: (icon: IconName) => string;

  /**
   * @param items The list of menu items to display.
   * @param header A header that can be used to separate from other block menus.
   */
  menu: (items: MenuItem[], header?: string) => void;
}

export interface Events {
  /**
   * Emit an Appsemble event.
   *
   * @param type The type of event to emit.
   * @param data Data to emit with the event.
   * @returns Boolean indicating whether an emitter is implemented or not.
   */
  emit: { [K in keyof EventEmitters]: (data: unknown, error?: string) => Promise<boolean> };

  /**
   * Remove an event listener for an Appsemble event.
   *
   * @param type The type of event to listen remove the listener from.
   * @param callback The callback to remove.
   * @returns Boolean indicating whether a listener is implemented or not.
   */
  off: {
    [K in keyof EventListeners]: <T>(callback: (data: T, error?: string) => void) => boolean;
  };

  /**
   * Add an event listener for an Appsemble event.
   *
   * @param type The type of event to listen on.
   * @param callback A callback to register for the event.
   * @returns Boolean indicating whether a listener is implemented or not.
   */
  on: {
    [K in keyof EventListeners]: <T>(callback: (data: T, error?: string) => void) => boolean;
  };
}

export interface EventParams {
  /**
   * The actions that may be dispatched by the block.
   */
  actions: { [K in keyof Actions]: Action };

  /**
   * Any kind of data that has been passed in by some context.
   */
  data: unknown;

  /**
   * Event related functions and constants.
   */
  events: Events;

  /**
   * Some utility functions provided by the Appsemble app framework.
   */
  utils: Utils;
}

/**
 * The parameters that get passed to the bootstrap function.
 */
export interface BootstrapParams extends EventParams {
  /**
   * URL parameters of the current route.
   *
   * If the page on which the block is rendered, has parameters specified on a page level, the
   * parameter keys and values of the page will be extracted and set as this object.
   */
  pageParameters?: PageParameters;

  /**
   * The path of the block as it is being rendered on the page using the page’s normalized name.
   */
  path: string;

  /**
   * The path of the block as it is being rendered on the page using the page’s index.
   */
  pathIndex: string;

  /**
   * The theme of the block given the context.
   *
   * This is based on a combination of the base theme, app theme, page theme and block theme.
   */
  theme: Theme;

  /**
   * The shadow root to which DOM elements may be appended.
   */
  shadowRoot: ShadowRoot;

  /**
   * Some utility functions provided by the Appsemble app framework.
   */
  utils: BlockUtils;

  /**
   * The parameters as they are defined in the app definition.
   */
  parameters: Parameters;
}

// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type BootstrapFunction = (params: BootstrapParams) => Promisable<Element | void>;

export interface AppsembleBootstrapEvent extends CustomEvent {
  detail: {
    fn: BootstrapFunction;
    document: Document;
  };
}

/**
 * Register a bootstrap function.
 *
 * @param fn The bootstrap function to register
 *
 *   If the function returns an element, it’s appended to the shadow root.
 */
export function bootstrap(fn: BootstrapFunction): void {
  const event: AppsembleBootstrapEvent = new CustomEvent('AppsembleBootstrap', {
    detail: {
      fn,
      document,
    },
  });
  if (document.currentScript) {
    document.currentScript.dispatchEvent(event);
  }
}

export type HandlerFunction = (...params: any[]) => Promisable<unknown>;

export type ControllerFunction = (
  params: EventParams,
) => Promisable<Record<keyof Actions, HandlerFunction>>;

export interface AppsembleControllerEvent extends CustomEvent {
  detail: {
    fn: ControllerFunction;
    document: Document;
  };
}

/**
 * Register a controller function.
 *
 * @param fn The controller function to register
 */
export function controller(fn: ControllerFunction): void {
  const event: AppsembleControllerEvent = new CustomEvent('AppsembleController', {
    detail: {
      fn,
      document,
    },
  });
  if (document.currentScript) {
    document.currentScript.dispatchEvent(event);
  }
}
