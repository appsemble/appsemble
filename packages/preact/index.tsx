import { type BootstrapParams, type Messages, bootstrap as sdkBootstrap } from '@appsemble/sdk';
import { type ComponentChild, type ComponentType, createContext, render, type VNode } from 'preact';
import { useContext } from 'preact/hooks';

export interface BlockProps extends BootstrapParams {
  /**
   * A function that must be called to indicate the block is ready to be rendered.
   */
  ready: () => void;
}

// eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error
// @ts-ignore 2345 argument of type is not assignable to parameter of type (strictNullChecks)
const Context = createContext<BlockProps>(null);

/**
 * Mount a Preact component returned by a bootstrap function in the shadow DOM of a block.
 *
 * @param Component The Preact component to mount.
 * @returns A promise which gets resolved if the component calls `ready()`.
 */
export function mount(
  Component: ComponentType<BlockProps>,
): (params: BootstrapParams) => Promise<void> {
  return (params) =>
    new Promise((ready) => {
      const props = {
        ...params,
        ready,
      };
      const component = (
        <Context.Provider value={props}>
          <Component {...props} />
        </Context.Provider>
      );
      render(component, params.shadowRoot);
      params.utils.addCleanup(() => render(null, params.shadowRoot));
    });
}

export function bootstrap(Component: ComponentType<BlockProps>): void {
  sdkBootstrap(mount(Component));
}

/**
 * A HOC which passes the Appsemble block values to he wrapped Preact component.
 *
 * @param Component The Preact component to wrap.
 * @returns The wrapper component.
 * @deprecated Use `useBlock()` instead.
 */
export function withBlock<P extends {}>(
  Component: ComponentType<Omit<BlockProps, keyof P> & P>,
): ComponentType<P> {
  return (props: P) => (
    <Context.Consumer>{(values) => <Component {...props} {...values} />}</Context.Consumer>
  );
}

export function useBlock(): BlockProps {
  return useContext(Context);
}

interface FormattedMessagePropsWithoutValues<M extends keyof Messages> {
  /**
   * The ID of the message to render.
   */
  id: M;
}

interface FormattedMessagePropsWithValues<M extends keyof Messages> {
  /**
   * The ID of the message to render.
   */
  id: M;

  /**
   * Values to pass to the formatted message.
   */
  values: Messages[M] extends never ? undefined : Messages[M];
}

/**
 * Render an Appsemble message.
 */
export function FormattedMessage<M extends keyof Messages>(
  props: Messages[M] extends never
    ? FormattedMessagePropsWithoutValues<M>
    : FormattedMessagePropsWithValues<M>,
): VNode {
  const { utils } = useBlock();

  // @ts-expect-error The messages interface isnâ€™t implemented within the Preact SDK.
  return utils.formatMessage(props.id, props.values) as ComponentChild as VNode;
}
