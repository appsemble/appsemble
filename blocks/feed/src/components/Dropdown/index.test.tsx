import { getDefaultBootstrapParams } from '@appsemble/block-interaction-tests';
import { Context } from '@appsemble/preact';
import { type BootstrapParams } from '@appsemble/sdk';
import { render, screen } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';
import { expect, it, vi } from 'vitest';

import { DropdownComponent } from './index.js';

type BlockProps = BootstrapParams & { ready: (value: PromiseLike<void> | void) => void };

const defaultBootstrapParams = getDefaultBootstrapParams();

it('should render a dropdown component', async () => {
  const onClickMock = vi.fn();

  const onClick = Object.assign(
    // eslint-disable-next-line func-names,prefer-arrow-callback
    function <R>(data?: any, context?: Record<string, any>): Promise<R> {
      return onClickMock(data, context) as Promise<R>;
    },
    { type: 'noop' as const },
  );

  const props: BlockProps = {
    ...defaultBootstrapParams,
    parameters: {
      dropdown: {
        icon: 'ellipsis-vertical',
        options: [{ onClick: 'onClick', label: 'Click' }],
      },
    },
    actions: {
      onClick,
      onAvatarClick: null,
      onButtonClick: null,
      onLoadReply: null,
      onSubmitReply: null,
    },
  };

  render(
    <Context.Provider value={props}>
      <DropdownComponent
        content={{ id: 1, status: '', photos: [] }}
        dropdown={props.parameters.dropdown}
      />
    </Context.Provider>,
  );

  const testDropdown = screen.getByTestId('dropdown');
  expect(testDropdown).toMatchSnapshot();

  await userEvent.click(testDropdown.getElementsByClassName('dropdown-trigger')[0]);
  await userEvent.click(testDropdown.getElementsByClassName('dropdown-item')[0]);
  expect(onClickMock).toHaveBeenCalledOnce();
});
