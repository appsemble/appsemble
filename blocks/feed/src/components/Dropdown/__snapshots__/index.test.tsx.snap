// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render a dropdown component 1`] = `
<div
  class="dropdown"
  data-testid="dropdown"
>
  <div
    class="dropdown-trigger"
  >
    <div
      aria-haspopup="true"
      role="button"
      tabindex="-1"
    >
      <span
        class="icon is-justify-content-end is-align-items-end"
      >
        <i
          class="ellipsis-vertical"
        />
      </span>
    </div>
  </div>
  <div
    class="dropdown-menu dropdown-menu"
    data-testid="dropdown-menu"
    role="menu"
    tabindex="0"
  >
    <div
      class="dropdown-content"
    >
      <button
        class="button dropdown-item pl-5 noBorder"
        type="button"
      >
        Click
      </button>
    </div>
  </div>
</div>
`;
